### PR TITLE
RHOAIENG-4924: Adding performance tuning section for single-model serving platform

### PR DESCRIPTION
**Pull request by syaseen-rh**
## Description
Adding performance tuning section for single-model serving platform

## How Has This Been Tested?
Local build

### CODE REVIEW

File: assemblies/serving-large-models.adoc [hunk 1]
```diff
@@ -25,6 +25,11 @@ endif::[]
 
 include::modules/viewing-metrics-for-the-single-model-serving-platform.adoc[leveloffset=+1]
 
+== Performance tuning on the single-model serving platform
+Certain performance issues may require you to tune the parameters of your inference service or model-serving runtime.
```
Comment (jbyrne-redhat) [inline assemblies/serving-large-models.adoc:29]: Use "may" only to indicate _permission_. Don't use it to indicate _possibility_. For that, use "might". 

See https://www.ibm.com/docs/en/ibm-style?topic=word-usage#may

File: modules/addressing-cuda-oom-errors.adoc [hunk 2]
```diff
@@ -0,0 +1,51 @@
+:_module-type: PROCEDURE
+
+[id="addressing-cuda-oom-errors-for-the-single-model-serving-platform_{context}"]
+= Addressing CUDA Out of Memory errors
```
Comment (jbyrne-redhat) [inline modules/addressing-cuda-oom-errors.adoc:4]: A couple of things:

- I suggest using "resolving" instead of "addressing". That's the verb we use in our release notes.
 - At Red Hat, we use sentence-style capitalization for headings, not headline-style. See https://redhat-documentation.github.io/supplementary-style-guide/#titles-and-headings. 
 - In this case, "out of memory" is a compound adjective that describes "errors". So it should all be hyphenated.
 - Putting that all together: "Resolving CUDA out-of memory errors"

File: modules/addressing-cuda-oom-errors.adoc [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+[role="_abstract"]
+
+In certain cases, depending on the model and hardware accelerator used, the TGIS memory auto tuning algorithm may underestimate the amount of GPU memory used to process long sequences. This can lead to CUDA OOM error responses from the model server. In such cases, we recommend updating or adding additional parameters in the TGIS model-serving runtime as described in the following procedure.
```
Comment (jbyrne-redhat) [inline modules/addressing-cuda-oom-errors.adoc:8]: A few things here:

- In this case, "auto tuning" is a compound adjective that describes "algorithm". Therefore, it should be  hyphenated to auto-tuning".
- Again, use "might" instead of "may".
- I think it should be "might underestimate the amount of GPU memory **needed**"?
- Don't use "this" without an object because it is ambiguous. In this case, I suggest "this miscalculation".
- The first use of an abbreviation in a module should spell out the word in full followed by the abbreviation in parentheses. Therefore, "Compute Unified Architecture (CUDA) out-of-memory (OOM) error responses".
- Never use "we". See https://www.ibm.com/docs/en/ibm-style?topic=word-usage#we. Instead, I suggest "To resolve these types of errors, you must update or add..."
- Add a comma before "as described in..."

File: modules/addressing-cuda-oom-errors.adoc [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+.Prerequisites
+* You have logged in to {productname-long}.
+ifdef::upstream[]
+* If you are using specialized {productname-short} groups, you are part of the admin group (for example, {odh-admin-group}) in {openshift-platform}.
+endif::[]
+ifndef::upstream[]
+* If you are using specialized {productname-short} groups, you are part of the admin group (for example, {oai-admin-group}) in {openshift-platform}.
+endif::[]
+* You have installed KServe.
```
Comment (jbyrne-redhat) [inline modules/addressing-cuda-oom-errors.adoc:19]: Should this prereq say "You have installed the single-model serving platform" and point to https://access.redhat.com/documentation/en-us/red_hat_openshift_ai_self-managed/2-latest/html/serving_models/serving-large-models_serving-large-models#about-the-single-model-serving-platform_serving-large-models?

File: modules/addressing-cuda-oom-errors.adoc [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+* You have enabled a custom model-serving runtime for TGIS.
```
Comment (jbyrne-redhat) [inline modules/addressing-cuda-oom-errors.adoc:20]: A couple of things here:

- Does it have to be a _custom_ runtime? Does this not apply to the TGIS runtime that we provide out of the box?
- Our documentation talks about **adding** custom runtimes more than enabling them (which happens by default). See https://access.redhat.com/documentation/en-us/red_hat_openshift_ai_self-managed/2-latest/html/serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models.

File: modules/addressing-cuda-oom-errors.adoc [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+
+.Procedure
+. From the {productname-short} dashboard, click *Settings* > *Serving runtimes*.
++
+The *Serving runtimes* page opens and shows the model-serving runtimes that are already installed and enabled.
++
+. Find your custom runtime for TGIS, click the action menu next to the runtime and select *Edit*.
```
Comment (jbyrne-redhat) [inline modules/addressing-cuda-oom-errors.adoc:28]: Again, wondering why we're talking about a custom TGIS runtime, when we provide one out of the box?

File: modules/addressing-cuda-oom-errors.adoc [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    env:
+    - name: BATCH_SAFETY_MARGIN
+      value: 30
+    - name: ESTIMATE_MEMORY_BATCH
+      value: 8
+----
++
+[NOTE]
+====
+The `BATCH_SAFETY_MARGIN` parameter sets a percentage of free GPU memory to hold back as a safety margin to avoid OOM conditions. The default value of `BATCH_SAFETY_MARGIN` is 20. The `ESTIMATE_MEMORY_BATCH_SIZE` parameter sets the batch size used in the memory auto-tuning algorithm. The default value of `ESTIMATE_MEMORY_BATCH_SIZE`  is 16.
```
Comment (jbyrne-redhat) [inline modules/addressing-cuda-oom-errors.adoc:47]: Enclose `20` and `16` in backticks (literal values).

File: modules/addressing-cuda-oom-errors.adoc [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+====
+. Click *Save*.
```
Comment (jbyrne-redhat) [inline modules/addressing-cuda-oom-errors.adoc:49]: What happens when you click Save? How does it affect models that are already running?

**Review by jbyrne-redhat — CHANGES_REQUESTED** [review]

File: modules/addressing-cuda-oom-errors.adoc (hunk 5 above)
Comment (syaseen-rh) [reply modules/addressing-cuda-oom-errors.adoc:20]: It doesn't look like there is currently an option to edit the default TGIS runtime. You can only edit custom runtimes. I will try to get further confirmation.

File: modules/addressing-cuda-oom-errors.adoc (hunk 6 above)
Comment (syaseen-rh) [reply modules/addressing-cuda-oom-errors.adoc:28]: Same comment as above.

File: modules/addressing-cuda-oom-errors.adoc (hunk 8 above)
Comment (syaseen-rh) [reply modules/addressing-cuda-oom-errors.adoc]: Addressed in the latest commit.

File: modules/resolving-cuda-oom-errors.adoc [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+endif::[]
+ifndef::upstream[]
+* If you are using specialized {productname-short} groups, you are part of the admin group (for example, {oai-admin-group}) in {openshift-platform}.
+* You have installed the single-model serving platform. For more information, see link:{rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#about-the-single-model-serving-platform_serving-large-models[About the single-model serving platform]
+* You have added a custom model-serving runtime for TGIS or have created a copy of the pre-installed standalone TGIS runtime. For more information, see link:{rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform].
```
Comment (jbyrne-redhat) [inline modules/resolving-cuda-oom-errors.adoc:21]: Hmmm, to me, these are more like steps in the core procedure than prerequisites?

I think I would add a conditional step (replacing the existing step 2 in the procedure) similar to the following:

```
. Based on the runtime that you used to deploy your model, perform one of the following actions:
+
* If you used the pre-installed *TGIS Standalone ServingRuntime for KServe* runtime, you must duplicate this runtime before following the remainder of this procedure. For more information about duplicating the pre-installed TGIS runtime, see.....
* If you were already using a custom TGI runtime, click the action menu (&#8942;) next to the runtime and select *Edit*.
```

File: modules/resolving-cuda-oom-errors.adoc [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
++
+[NOTE]
+====
+The `BATCH_SAFETY_MARGIN` parameter sets a percentage of free GPU memory to hold back as a safety margin to avoid OOM conditions. The default value of `BATCH_SAFETY_MARGIN` is `20`. The `ESTIMATE_MEMORY_BATCH_SIZE` parameter sets the batch size used in the memory auto-tuning algorithm. The default value of `ESTIMATE_MEMORY_BATCH_SIZE`  is `16`.
+====
+. Click *Update*.
++ 
+The *Serving runtimes* page opens and shows the list of runtimes that are installed. Observe that the custom model-serving runtime you updated is shown.
++
+. You must redeploy the impacted model for the updated or added parameters in the model-serving runtime to take effect:
```
Comment (jbyrne-redhat) [inline modules/resolving-cuda-oom-errors.adoc:54]: I'd rephrase this to put the condition first:

`To redeploy the model for the parameter updates to take effect, perform the following actions:`

File: modules/resolving-cuda-oom-errors.adoc [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+.. From the {productname-short} dashboard, click *Model Serving* > *Deployed Models*.
+.. Find the model you want to update, click the action menu (⋮) beside the model, and click *Delete*.
```
Comment (jbyrne-redhat) [inline modules/resolving-cuda-oom-errors.adoc:56]: A couple of suggestions here:

- "Find the model **that** you want to **redeploy**..."
- "click the action menu (&#8942;) next to the model..."

- click the action menu (&#8942;) next to the model...

File: modules/resolving-cuda-oom-errors.adoc [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+ifndef::upstream[]
+.. Redeploy the model as described link:{rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#deploying-models-on-the-single-model-serving-platform_serving-large-models[here].
```
Comment (jbyrne-redhat) [inline modules/resolving-cuda-oom-errors.adoc:58]: Never use "here" for a link. Use the proper title of the section.

File: modules/resolving-cuda-oom-errors.adoc [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+endif::[]
+ifdef::upstream[]
+.. Redeploy the model as described link:{odhdocshome}/serving_models/#deploying-models-on-the-single-model-serving-platform_serving-large-models[here].
```
Comment (jbyrne-redhat) [inline modules/resolving-cuda-oom-errors.adoc:61]: Ditto

File: modules/resolving-cuda-oom-errors.adoc [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+ifdef::upstream[]
+* If you are using specialized {productname-short} groups, you are part of the admin group (for example, {odh-admin-group}) in {openshift-platform}.
+* You have added a custom model-serving runtime for TGIS or have created a copy of the pre-installed standalone TGIS runtime. For more information, see link:{odhdocshome}/serving_models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform].
+endif::[]
+ifndef::upstream[]
+* If you are using specialized {productname-short} groups, you are part of the admin group (for example, {oai-admin-group}) in {openshift-platform}.
+* You have added a custom model-serving runtime for TGIS or have created a copy of the pre-installed standalone *TGIS Standalone ServingRuntime for KServe* runtime. For more information, see link:{rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform].
```
Comment (jbyrne-redhat) [inline modules/resolving-cuda-oom-errors.adoc:19]: Remove this prereq now that the information is in the core procedure.

File: modules/resolving-cuda-oom-errors.adoc [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+endif::[]
+
+.Procedure
+. From the {productname-short} dashboard, click *Settings* > *Serving runtimes*.
++
+The *Serving runtimes* page opens and shows the model-serving runtimes that are already installed and enabled.
++
+. Based on the runtime that you used to deploy your model, perform one of the following actions:
++
+* If you used the pre-installed *TGIS Standalone ServingRuntime for KServe* runtime, you must duplicate this runtime before following the remainder of this procedure. For more information about duplicating the pre-installed TGIS runtime, see {rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform].
```
Comment (jbyrne-redhat) [inline modules/resolving-cuda-oom-errors.adoc:27]: A few tweaks to my original suggested wording (sorry!):

```
If you used the pre-installed *TGIS Standalone ServingRuntime for KServe* runtime, duplicate the runtime to create a custom version and then follow the remainder of this procedure. To learn how to duplicate the pre-installed TGIS runtime, see {rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform].
```

Comment (syaseen-rh) [reply modules/resolving-cuda-oom-errors.adoc]: Good suggestion! Updated in latest commit.

**Review by jbyrne-redhat — APPROVED** [review]
